tests/components/iaqualink/test_init.py

```diff
@@ -0,0 +1,148 @@
+        await hass.async_block_till_done()
+
+    mock_setup.assert_not_called()
+
+
+async def test_setup_login_exception(hass):
+    """..."""
+    entry = MOCK_CONFIG
+    entry.add_to_hass(hass)
+
+    with patch(
+        "iaqualink.client.AqualinkClient.login",
+        side_effect=iaqualink.exception.AqualinkServiceException,
+    ):
+        assert not await ha_iaqualink.async_setup_entry(hass, entry)
+
+
+async def test_setup_login_timeout(hass):
+    """..."""
+    entry = MOCK_CONFIG
+    entry.add_to_hass(hass)
+
+    with patch(
+        "iaqualink.client.AqualinkClient.login",
+        side_effect=asyncio.TimeoutError,
+    ), pytest.raises(ConfigEntryNotReady):
+        await ha_iaqualink.async_setup_entry(hass, entry)
+
+
+async def test_setup_systems_exception(hass):
+    """..."""
+    entry = MOCK_CONFIG
+    entry.add_to_hass(hass)
+
+    with patch("iaqualink.client.AqualinkClient.login", return_value=None), patch(
+        "iaqualink.client.AqualinkClient.get_systems",
+        side_effect=iaqualink.exception.AqualinkServiceException,
+    ), pytest.raises(ConfigEntryNotReady):
+        await ha_iaqualink.async_setup_entry(hass, entry)
+
+
+async def test_setup_devices_exception(hass):
+    """..."""
+    entry = MOCK_CONFIG
+    entry.add_to_hass(hass)
+
+    with patch("iaqualink.client.AqualinkClient.login", return_value=None), patch(
+        "iaqualink.client.AqualinkClient.get_systems", return_value=MOCK_SYSTEMS
+    ), patch(
+        "iaqualink.system.AqualinkSystem.get_devices",
+        side_effect=iaqualink.exception.AqualinkServiceException,
+    ), pytest.raises(
+        ConfigEntryNotReady
+    ):
+        await ha_iaqualink.async_setup_entry(hass, entry)
+
+
+async def test_setup_all_good_no_recognized_devices(hass):
+    """..."""
+    entry = MOCK_CONFIG
+    entry.add_to_hass(hass)
+
+    with patch("iaqualink.client.AqualinkClient.login", return_value=None), patch(
+        "iaqualink.client.AqualinkClient.get_systems", return_value=MOCK_SYSTEMS
+    ), patch(
+        "iaqualink.system.AqualinkSystem.get_devices", return_value=MOCK_UNKNOWN_DEVICES
+    ):
+        assert await ha_iaqualink.async_setup_entry(hass, entry)
+
+    assert hass.data[ha_iaqualink.DOMAIN][BINARY_SENSOR_DOMAIN] == []
+    assert hass.data[ha_iaqualink.DOMAIN][CLIMATE_DOMAIN] == []
+    assert hass.data[ha_iaqualink.DOMAIN][LIGHT_DOMAIN] == []
+    assert hass.data[ha_iaqualink.DOMAIN][SENSOR_DOMAIN] == []
+    assert hass.data[ha_iaqualink.DOMAIN][SWITCH_DOMAIN] == []
+
+
+async def test_setup_all_good_all_device_types(hass):
+    """..."""
+    entry = MOCK_CONFIG
+    entry.add_to_hass(hass)
+
+    with patch("iaqualink.client.AqualinkClient.login", return_value=None), patch(
+        "iaqualink.client.AqualinkClient.get_systems", return_value=MOCK_SYSTEMS
+    ), patch("iaqualink.system.AqualinkSystem.get_devices", return_value=MOCK_DEVICES):
+        assert await ha_iaqualink.async_setup_entry(hass, entry)
```

Comment:
Please use the standard interface instead of calling `async_setup_entry` directly
https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations